### PR TITLE
Assert input shapes of Merge layer are equal

### PR DESCRIPTION
**Pull request by farizrahman4u**
Make sure that only layers with same output_shape are merged using sum, ave and mul modes.

### CODE REVIEW

**Comment by EderSantana** [comment]
Won't theano already catch this error?

**Comment by farizrahman4u** [comment]
@EderSantana  Of course there will be some error while training. But no exception is thrown during creation of the Merge layer or even while compiling. I tried the following model and it compiles successfully:

```
>>>python
>>> import keras
>>> from keras.models import Sequential
>>> from keras.layers.core import Dense, Merge
>>> left=Sequential()
>>> left.add(Dense(input_dim=10,output_dim=20))
>>> right=Sequential()
>>> right.add(Dense(input_dim=10,output_dim=15))
>>> model=Sequential()
>>> model.add(Merge([left,right],'sum'))
>>> model.add(Dense(1))
>>> model.compile(loss='mse',optimizer='sgd')
```

**Comment by EderSantana** [comment]
I see. Bad think about scripting languages is that all type checking has to be manual, right?

But why only the `merge` test is failing?

**Comment by farizrahman4u** [comment]
Not sure why.

**Comment by farizrahman4u** [comment]
@EderSantana Any idea what is wrong ?

**Comment by farizrahman4u** [comment]
Let me try restarting the build.

**Comment by farizrahman4u** [comment]
I think its Merge ready now.

File: keras/layers/core.py [hunk 1]
```diff
@@ -269,6 +271,10 @@ def __init__(self, layers, mode='sum', concat_axis=-1):
             raise Exception("Please specify two or more input layers (or containers) to merge")
         if mode not in {'sum', 'mul', 'concat', 'ave', 'join'}:
             raise Exception("Invalid merge mode: " + str(mode))
+        if mode in {'sum', 'mul', 'ave'}:
+            input_shapes=set([l.output_shape for l in layers])
+            if len(input_shapes)>1:
```
Comment (fchollet) [inline keras/layers/core.py]: Spaces around `<`

File: keras/layers/core.py [hunk 2]
```diff
@@ -79,8 +79,10 @@ def set_input_shape(self, input_shape):
     @property
     def output_shape(self):
         # default assumption: tensor shape unchanged.
-        return self.input_shape
-
+        try:
```
Comment (fchollet) [inline keras/layers/core.py]: Let's no add this try/catch, it is dangerous and confusing.

**Comment by fchollet** [comment]
Cool, we can definitely add this check! But then for consistency you would have to add a similar check for mode `concat`: check that all tensors have the same shape except for the dimension specified by `concat_axis`.

File: keras/layers/core.py (hunk 1 above)
Comment (fchollet) [inline keras/layers/core.py]: Also around `=`

**Comment by farizrahman4u** [comment]
@fchollet Done!

**Comment by dbonadiman** [comment]
That screwed up my code, no more concat merges on Embeddings. I had to remove the shape check for concat to get it work.

**Pull request merged**